### PR TITLE
Changed GapAdvertisingData.addData to take current fields into account when adding data.

### PR DESCRIPTION
**Pull request by marcuschangarm**
Resubmission of https://github.com/ARMmbed/ble/pull/101

### CODE REVIEW

File: ble/GapAdvertisingData.h [hunk 1]
```diff
@@ -210,27 +211,95 @@ class GapAdvertisingData
...
+                    if (len == dataLength) {
+                        for (uint8_t idx = 0; idx < dataLength; idx++) {
+                            field[2 + idx] = payload[idx];
+                        }
+                    } else {
+                        // check if data fits
+                        if ((_payloadLen - dataLength + len) <= GAP_ADVERTISING_DATA_MAX_PAYLOAD) {
+                            // remove old field
+                            while (field + dataLength < &_payload[GAP_ADVERTISING_DATA_MAX_PAYLOAD]) {
+                                *field = field[dataLength];
```
Comment (rgrover) [inline ble/GapAdvertisingData.h]: shouldn't this be `*field = field[dataLength +1];`?

File: ble/GapAdvertisingData.h [hunk 2]
```diff
@@ -210,27 +211,95 @@ class GapAdvertisingData
...
+                    // new data has same length, do in-order replacement
+                    if (len == dataLength) {
+                        for (uint8_t idx = 0; idx < dataLength; idx++) {
+                            field[2 + idx] = payload[idx];
+                        }
+                    } else {
+                        // check if data fits
+                        if ((_payloadLen - dataLength + len) <= GAP_ADVERTISING_DATA_MAX_PAYLOAD) {
+                            // remove old field
+                            while (field + dataLength < &_payload[GAP_ADVERTISING_DATA_MAX_PAYLOAD]) {
```
Comment (rgrover) [inline ble/GapAdvertisingData.h]: shouldn't this be `while ((field + dataLength + 1) < &_payload[GAP_ADVERTISING_DATA_MAX_PAYLOAD]) {`?

File: ble/GapAdvertisingData.h [hunk 3]
```diff
@@ -210,27 +211,95 @@ class GapAdvertisingData
...
+                        field[1] = advDataType;
+
+                        result = BLE_ERROR_NONE;
+                    }
+
+                    break;
+                }
+                // Field exists but is not recognized. Abort operation.
+                default:
+                    break;
```
Comment (rgrover) [inline ble/GapAdvertisingData.h]: `result` could be set to a more useful value here.

File: ble/GapAdvertisingData.h [hunk 4]
```diff
@@ -210,27 +211,95 @@ class GapAdvertisingData
...
+                        for (uint8_t idx = 0; idx < len; idx++) {
+                            field[2 + idx] = payload[idx];
+                        }
+
+                        // increment lengths
+                        field[0] += len;
+                        _payloadLen += len;
+
+                        // re-insert TYPE field
+                        field[1] = advDataType;
```
Comment (rgrover) [inline ble/GapAdvertisingData.h]: this is not necessary, is it?

**Comment by marcuschangarm** [comment]
It was actually an off-by-2-error, dataLength + type field + length field. And I forgot to decrement _payloadLen.

I changed result to BLE_ERROR_NOT_IMPLEMENTED if the user tries to update a field that is not supported.

**Pull request merged**